### PR TITLE
feat: Added generatePackageTree 

### PR DESCRIPTION
**Pull request by jsamar17**
generatePackageTree takes in a root directory of a package tree and returns a package tree with the project as the head and populated with its dependencies

### CODE REVIEW

File: src/package-tree.ts [hunk 1]
```diff
@@ -1,53 +1,110 @@
...
 }
 
-function getPOIforPackageTree(packageTree: PackageTree): PackageTree {
-  // TODO:
-  throw new Error('not implemented');
-  // step 1: get POI modules for current package
-  // step 2: add POI to PackageTree Object
-}
+/**
+ * Takes in the contents of a project's package.json file as a string and returns a PackageTree ob
```
Comment (jplaisted) [inline src/package-tree.ts]: > ob

...ject?

File: src/package-tree.ts [hunk 2]
```diff
@@ -1,53 +1,110 @@
...
+ * @param pjson is the contents of a project's package.json file as a string
+ */
+export async function generatePackageTree(rootDir: string, customReadFilep: ReadFileP): Promise<PackageTree<undefined>> {
+  // todo -- this function will probably be recursive
+  customReadFilep = customReadFilep || readFilep;
+  // step 0: read in package.json and package-lock.json
+  try{
```
Comment (jplaisted) [inline src/package-tree.ts]: Run clang format. Needs to be a space after the `try`, and the this entire block needs to be indented.

File: src/package-tree.ts [hunk 3]
```diff
@@ -1,53 +1,110 @@
...
+  const pjsonPath = path.join(rootDir, 'package.json');
```
Comment (jplaisted) [inline src/package-tree.ts]: Why do these variables have `p` prefixes?

File: src/package-tree.ts [hunk 4]
```diff
@@ -1,53 +1,110 @@
...
 
-// Gets Points of Interest for a single file
-function getPointsOfInterest(
-    contents: string, fileName: string,
-    functionArray: POIDetector[]): PointOfInterest[] {
-  // TODO:
-  throw new Error('not implemented');
+/**
+ * Takes in a list of depndencies and returns a constructed PackageTree obj
+ * @param dependencies
```
Comment (jplaisted) [inline src/package-tree.ts]: No need to have `@param` unless you have something to say about that parameter.

File: src/package-tree.ts [hunk 5]
```diff
@@ -1,53 +1,110 @@
...
 }
 
-// Gets Points of Interest for a single file
-function getPointsOfInterest(
-    contents: string, fileName: string,
-    functionArray: POIDetector[]): PointOfInterest[] {
-  // TODO:
-  throw new Error('not implemented');
+/**
+ * Takes in a list of depndencies and returns a constructed PackageTree obj
```
Comment (jplaisted) [inline src/package-tree.ts]: > obj

...ect?

File: src/package-tree.ts [hunk 6]
```diff
@@ -1,53 +1,116 @@
...
-function getPointsOfInterest(
-    contents: string, fileName: string,
-    functionArray: POIDetector[]): PointOfInterest[] {
-  // TODO:
-  throw new Error('not implemented');
+// Step 2: For each Package Tree obj get the dependencies
+const m: Array<Promise<void>> = packageTreeArr.map(async element => {
+  element.dependencies = await populateDependencies(element, packageLock);
+});
+await Promise.all(m);
```
Comment (kjin) [inline src/package-tree.ts]: Given that `m` isn't used anywhere else, you can just `await` the RHS of the previous statement.

File: src/package-tree.ts (hunk 1 above)
Comment (jsamar17) [reply src/package-tree.ts]: fixed!

File: src/package-tree.ts [hunk 7]
```diff
@@ -1,53 +1,116 @@
...
+ const dependencies = packageLock.dependencies[packageName].requires;
+ if(!dependencies){
+   return undefined;
+ }
+ 
+ const dependencyPackageTree = await getPackageTreeFromDependencyList(dependencies, packageLock);
+ if(dependencyPackageTree){
+   return dependencyPackageTree;
+ }
+ return undefined;
```
Comment (kjin) [inline src/package-tree.ts]: Same note here concerning [] vs undefined. You can shorthand the last four lines as `return dependencyPackageTree || undefined`.

File: test/test-package-tree.ts [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+                    'data': undefined,
+                    'dependencies': []
+                }]
+            }
+        ];
+        t.deepEqual(result, expectedResult);
+    });
+
+    test(
+        'getPackageTreeFromDependencyList should return a populated packageTree array',
```
Comment (kjin) [inline test/test-package-tree.ts]: The test message should uniquely identify the behavior being tested. This looks like the same string as in the previous test. How can you re-phrase this to show that the individual behavior being tested at a glance?

File: src/package-tree.ts (hunk 3 above)
Comment (jsamar17) [reply src/package-tree.ts]: they are objects representing package.json and package-lock.json. I edited the names though

File: src/package-tree.ts (hunk 6 above)
Comment (jplaisted) [reply src/package-tree.ts]: He still needs the `Promise.all`, which makes that previous line that much longer. I'm fine either way, but I see why he made the `m`.

Comment (jsamar17) [reply src/package-tree.ts]: I came to the same conclusion, when I try to await the RHS my tests fail. So I am going to keep it unless there is a more readable way to write it out that is suggested

Comment (jsamar17) [reply src/package-tree.ts]: I changed it so Promise.all is called on the return value of packageTreeArr.map

File: src/package-tree.ts [hunk 9]
```diff
@@ -194,3 +236,65 @@ function main() {
...
+export async function getPackageTreeFromDependencyList(
+    dependencies: {},
+    packageLockJson: {}): Promise<Array<PackageTree<undefined>>> {
+  // Step 1: For each dependency create a PackageTree obj with the name and
+  // version fields populated
+  if (!dependencies) {
+    return [];
+  }
+
+  const dependencyArr: Array<[string, string]> = Object.entries(dependencies);
```
Comment (kjin) [inline src/package-tree.ts]: Did you end up changing `Object.entries`?

File: src/package-tree.ts [hunk 10]
```diff
@@ -23,11 +31,45 @@ export interface PackageTree<T> {
...
+    const treeHead: PackageTree<undefined> = {
+      name: packageJson.name,
+      version: packageJson.version,
+      data: undefined,
+      dependencies: result
+    };
+
+    return treeHead;
+  } catch (error) {
+    throw error;
```
Comment (kjin) [inline src/package-tree.ts]: Do we necessarily need to re-throw here?

Comment (jsamar17) [reply src/package-tree.ts]: No, I removed this

File: src/package-tree.ts (hunk 9 above)
Comment (jsamar17) [reply src/package-tree.ts]: Yes, I added it in tsconfig.json

File: src/package-tree.ts [hunk 11]
```diff
@@ -64,12 +63,9 @@ export async function generatePackageTree(
       version: packageJson.version,
       data: undefined,
       dependencies: result
-    };
+    }
```
Comment (kjin) [inline src/package-tree.ts]: I think `gts fix` will probably put this semicolon back.

**Pull request merged**